### PR TITLE
Fix cholesky_inverse JVP

### LINKED ISSUE #196682: torch.cholesky_inverse returns an incorrect JVP for a valid Cholesky factor
### 🐛 Describe the bug

`torch.cholesky_inverse` returns `-0.4988266424013123` for a first-order JVP whose mathematically correct value is `-0.3358436800325667` in `float64`.

The input is an invertible lower-triangular factor with positive diagonal. Writing `a = t + 1`, the sum of `(L L.T)**-1` is `5/(4*a**2) - 1/(2*a) + 1/4`. Differentiating that explicit inverse at `t = 0.7` gives `-0.3358436800325667`. PyTorch’s forward value matches the reference, but the JVP does not.

This issue reproduces on CPU, so CUDA availability does not affect the result.

```python
import torch

dtype = torch.float64

def f(t):
    one = torch.ones((), dtype=t.dtype)
    zero = torch.zeros((), dtype=t.dtype)
    L = torch.stack((t + one, zero, one, 2 * one)).reshape(2, 2)
    return torch.cholesky_inverse(L, upper=False).sum()

t = torch.tensor(0.7, dtype=dtype)
_, actual = torch.func.jvp(f, (t,), (torch.ones_like(t),))
expected = torch.tensor(-0.3358436800325667, dtype=dtype)

print("actual:")
print(actual)
print("expected:")
print(expected)
```

### Actual vs expected result

```text
actual:
-0.4988266424013123

expected:
-0.3358436800325667
```

The discrepancy occurs in the first-order forward-mode derivative. The forward value is `0.3884083044982699` and matches the explicit inverse; the JVP is wrong.

### Versions

```text
PyTorch version: 2.12.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
Python version: 3.13.13 | packaged by Anaconda, Inc.
Python platform: Linux x86_64

Is CUDA available: False
GPU models:
  GPU 0: NVIDIA RTX 6000 Ada Generation
  GPU 1: NVIDIA RTX 6000 Ada Generation
Nvidia driver version: 570.211.01

CPU:
  Model name: AMD Ryzen Threadripper PRO 7985WX 64-Cores
  CPU(s): 128

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[pip3] triton==3.7.0
```

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano

### PR DESCRIPTION
**Pull request by yinli-systems**
Fixes #196682.

The existing JVP places the inverse between the tangent and the Cholesky factor. Since matrix multiplication is noncommutative, this produces an incorrect tangent for non-diagonal factors.

This changes the partial derivative from `-X dF X F` to `-X dF F X`, preserving the existing conjugate-transpose handling for lower and upper factors. It also restores the derivation with the corrected final step.

The regression coverage lives in the existing OpInfo sample generator. A deterministic non-diagonal factor makes the operand-ordering error visible to forward-mode gradcheck for real and complex dtypes and for lower and upper factors. The sample is gated to `cholesky_inverse`, because `cholesky_solve` reuses this generator but has a separate pre-existing factor-derivative issue for non-diagonal inputs.

Tests:
- Focused CPU forward-mode and forward-over-reverse OpInfo tests for `cholesky_inverse` and `cholesky_solve`: 8 passed, 4 skipped
- `python -m spin quicklint`
- `lintrunner -a`

All executed tests and linters pass on macOS arm64 CPU.

AI assistance was used in preparing the implementation and analysis. I reviewed the final code, derivation, tests, and evidence and take responsibility for the contribution.

### CODE REVIEW

**Comment by Skylion007** [comment]
@claude review this PR

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -2326,14 +2326,11 @@ Tensor cholesky_inverse_backward(
 }
 
 // If X = (L L^H)^{-1} with L lower-triangular with a real positive diagonal,
-// then dX = K^H + K, where
-// K =  L^{-H} dL^{-1} [dL^{-1} = -L^{-1} dL L^{-1}]
-//   = -L^{-H} L^{-1} dL L^{-1} [L^{-H} L^{-1} = X]
-//   = -X dL L^{-1} [X = X^H = L^{-H} L^{-1} = L^{-1} L^{-H}]
-//   = -X dL X L^{H}.
+// then dX = -X (dL L^H + L dL^H) X = K + K^H, where
+// K = -X dL L^H X.
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp]: The math checks. Let's keep the derivation though, with the fixed last step `-X dL L^-1 = -X dL L^H (L^-H L^-1) = -X dL L^H X`.

**Comment by nikitaved** [comment]
How did it pass the tests, I wonder?

File: test/test_linalg.py [hunk 2]
```diff
@@ -8788,6 +8788,71 @@ def run_test(shape, batch, upper, contiguous):
...
 
+    @skipCUDAIfNoCusolver
+    @skipCPUIfNoLapack
+    @dtypes(torch.double, torch.cdouble)
+    @parametrize("upper", [False, True])
+    @parametrize("batched", [False, True])
+    @parametrize("noncontiguous", [False, True])
+    def test_cholesky_inverse_jvp(
+        self, device, dtype, upper, batched, noncontiguous
+    ):
```
Comment (nikitaved) [inline test/test_linalg.py]: Can't we just update OpInfo samples? Seems like our samples are insufficient.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (yinli-systems) [reply torch/csrc/autograd/FunctionsManual.cpp]: I reviewed this change and restored the derivation with the corrected final step in `9a7a88acaa`.

> AI-assisted implementation, reviewed by me: the lower-triangular derivation now keeps `-X dL L^{-1} = -X dL L^H (L^{-H} L^{-1}) = -X dL L^H X`, while the upper-triangular formula remains unchanged.

File: test/test_linalg.py (hunk 2 above)
Comment (yinli-systems) [reply test/test_linalg.py]: Agreed. I reviewed the updated test strategy and moved the regression into the OpInfo sample generator in `9a7a88acaa`, using a deterministic non-diagonal factor for real/complex and lower/upper coverage; the standalone `test_linalg.py` test is removed.

> AI-assisted validation summary, reviewed by me: the old `cholesky_inverse` JVP fails forward-mode gradcheck on this sample, while the corrected formula matches the reference JVP for real/complex and lower/upper cases (maximum observed error about `1.4e-16`). I did not pass the stronger factor into `cholesky_solve`: both its forward-mode and backward gradchecks currently fail with respect to a non-diagonal factor for 2x2 and 3x3 inputs on PyTorch 2.8, which is a separate pre-existing derivative issue. The shared generator therefore has an explicit opt-out and TODO for `cholesky_solve`. `spin quicklint` passes.

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 3]
```diff
@@ -593,7 +593,13 @@ def np_vander_batched(x, N=None):
...
 
 def sample_inputs_linalg_cholesky_inverse(
-    op_info, device, dtype, requires_grad=False, **kwargs
+    op_info,
+    device,
+    dtype,
+    requires_grad=False,
+    *,
+    include_nontrivial_factors=True,
+    **kwargs,
```
Comment (nikitaved) [inline torch/testing/_internal/opinfo/definitions/linalg.py]: No need to include additional params.

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 4]
```diff
@@ -613,7 +619,20 @@ def sample_inputs_linalg_cholesky_inverse(
...
+            dtype=dtype,
+            device=device,
+        )
+        if dtype.is_complex:
+            nontrivial_factor = nontrivial_factor + 1j * torch.tensor(
+                [[0.0, 0.0, 0.0], [0.25, 0.0, 0.0], [-0.5, 0.125, 0.0]],
+                dtype=dtype,
+                device=device,
+            )
+        test_cases.append(nontrivial_factor)
```
Comment (nikitaved) [inline torch/testing/_internal/opinfo/definitions/linalg.py]: Let's just append then and provide some context. Providing references to the issue/PR is also helpful.

**Comment by yinli-systems** [comment]
I reviewed the latest diff and confirmed that `2ee7159140` addresses both follow-ups: the shared generator signature is unchanged, and the non-diagonal factor is appended only for `cholesky_inverse` with issue context. I also updated the PR description so it reflects the OpInfo-based regression coverage rather than the removed standalone test.

> AI-assisted validation summary, reviewed by me: `python -m spin quicklint` and `lintrunner -a` pass on the current head. The two addressed review threads are now resolved.

@nikitaved, could you take another look when convenient?

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 5]
```diff
@@ -613,7 +613,23 @@ def sample_inputs_linalg_cholesky_inverse(
...
     )
-    test_cases = (torch.linalg.cholesky(a, upper=False) for a in inputs)
+    test_cases = [torch.linalg.cholesky(a, upper=False) for a in inputs]
+    if op_info.name == "cholesky_inverse":
+        # Regression sample for https://github.com/pytorch/pytorch/issues/196682.
+        # Keep it specific to cholesky_inverse: cholesky_solve also reuses this
+        # generator, but its factor derivative fails for non-diagonal inputs.
+        nontrivial_factor = torch.tensor(
+            [[2.0, 0.0, 0.0], [0.5, 1.5, 0.0], [-0.25, 0.75, 1.25]],
+            dtype=dtype,
```
Comment (nikitaved) [inline torch/testing/_internal/opinfo/definitions/linalg.py:625]: Let's add a note why this sample is special, for future references.

Comment (yinli-systems) [reply torch/testing/_internal/opinfo/definitions/linalg.py:625]: Addressed in `1ba98cc46a`. The comment now explains that the existing well-conditioned samples yield nearly diagonal factors, which can hide an operand-ordering error because the relevant products almost commute; the added sample uses order-one off-diagonal entries to expose the noncommuting JVP products.

I reviewed this wording and the surrounding generator logic. Validation for this comment-only change: `git diff --check` and Python syntax compilation pass.

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 6]
```diff
@@ -616,6 +616,8 @@ def sample_inputs_linalg_cholesky_inverse(
     test_cases = [torch.linalg.cholesky(a, upper=False) for a in inputs]
     if op_info.name == "cholesky_inverse":
         # Regression sample for https://github.com/pytorch/pytorch/issues/196682.
+        # Unlike the nearly diagonal factors above, its order-one off-diagonal
+        # entries expose the noncommuting matrix products in the JVP formula.
```
Comment (nikitaved) [inline torch/testing/_internal/opinfo/definitions/linalg.py:620]: Sorry, what do you mean by nearly diagonal? I am not sure this statement is correct -- the samples include pd matrices which are certainly not near diagonal. Maybe just scaling some sample up could be sufficient? I do not see why this specific sample breaks things -- that is the comment I meant.

Comment (yinli-systems) [reply torch/testing/_internal/opinfo/definitions/linalg.py]: You're right that my earlier wording was ambiguous: I meant the **Cholesky factors produced by this generator**, not that every positive-definite matrix is nearly diagonal.

`random_well_conditioned_matrix` uses `mean=1.0, sigma=0.001` for its singular values. Therefore `M @ M.mH = U diag(s²) Uᴴ` is close to identity, and its Cholesky factor makes `X` and `L.mH` nearly commute. That suppresses the difference between the old `-X @ dL @ X @ L.mH` order and the correct `-X @ dL @ L.mH @ X` order.

I checked this directly in float64. A default-style sample had normalized `||X Lᴴ - Lᴴ X|| / (||X|| ||L||) = 6.2e-7` and the old native JVP differed from the reference by about `1.2e-6` relatively. The added factor gives `0.140` and `0.382`, respectively, so it reliably exposes the operand-order bug.

Uniform scaling is not sufficient: with `L -> cL` (and the same tangent), both the correct JVP and the old/correct difference scale as `c^-3`, so their relative discrepancy is unchanged. I verified that at `c = 0.1, 1, 10` (all about `0.3815`).

I updated the comment in `989b13bbe1` to state this precisely and avoid the misleading “nearly diagonal” shorthand. Thanks for pushing on it.

Comment (nikitaved) [reply torch/testing/_internal/opinfo/definitions/linalg.py]: I am sorry, how is "Therefore M @ M.mH = U diag(s²) Uᴴ is close to identity" true? U is othonormal but not an identity. I believe the argument is more involved (operator norm within the same very narrow interval for both products thanks to the sampled singular values -- that I agree with), but I see why scaling is not sufficient with this distribution. Let's just fix the argument, and write that the new L factor implies a wider spectral margin to distinguish two products?

Comment (nikitaved) [reply torch/testing/_internal/opinfo/definitions/linalg.py]: We also have a method that samples a matrix with distinct singular values with a condition number of up to 3 -- maybe that would suffice? Then there is no need in explaining much. And that would fix the whole test as well. Let's fix it once and for all?

Comment (yinli-systems) [reply torch/testing/_internal/opinfo/definitions/linalg.py]: Yes, this is cleaner. I pushed `768ecb27f7`, replacing the hand-constructed factor and the long explanation with the existing `make_fullrank_matrices_with_distinct_singular_values` helper. The regression sample now forms `M @ M.mH` and takes its Cholesky factor.

I also checked that this is not merely a nicer-looking sample: across 50 seeds in float64 and complex128, for both upper and lower forms, the old JVP's relative discrepancy from the independent differential was at least 0.16 (median 0.31–0.36). So the helper's spectral spread robustly exposes the matrix-order error. PYFMT, Ruff, `py_compile`, and `git diff --check` pass locally.
